### PR TITLE
[MOSTLY MODULAR][DRAFT] NIFs

### PR DESCRIPTION
**Pull request by softcerv**
## About The Pull Request
https://hackmd.io/kX-LAehRTT2Cdgvu9qAJSw (I still need to update this design document more)

This is a remake of NIFs from Virgo. 

_**Nothing that has been made so far has been set in stone. This PR is a draft.**_

## Things that need to be done.

- [x] Implement alternative power drain methods
- [ ] Program in NIFSofts
- [ ] Add in custom examine text for NIFSofts
- [ ] Move NIF examine text to preference
- [ ] Add in NIF calibration
- [ ] Refine the NIFSoft vendor
- [ ] Add in custom lore.
- [ ] Add in more NIF models
- [ ] Add in custom textures for both the implants and the actions.
- [ ] Add more custom themes.

## What has been done.

- The actual NIF implants
- A working menu system with themes and flavor text
- The basis for NIFSofts
- A rudimentary NIFSoft vendor
-  Round to Round persistence

## Changelog

:cl:
add: adds in NIFs
/:cl:

### CODE REVIEW

**Comment by YakumoChen** [comment]
image
This is probably my only concern, given the uses both as an antag and against an antag. Could you elaborate on your plans for this, and how usable this is in antagonist situations? I'm not keen on giving users a potentially silent, undetectable, unstrippable call for help.

**Comment by softcerv** [comment]
> image This is probably my only concern, given the uses both as an antag and against an antag. Could you elaborate on your plans for this, and how usable this is in antagonist situations? I'm not keen on giving users a potentially silent, undetectable, unstrippable call for help.

Lets talk about this in the NIF thread under development-code. None of the stuff that I set out in this design doc is really set out in stone.

File: modular_skyrat/modules/modular_implants/code/nifs.dm [hunk 1]
```diff
@@ -0,0 +1,242 @@
+/// This is the original NIF that other NIFs are based on.
+/obj/item/organ/internal/cyberimp/brain/nif
+	name = "Nanite Implant Frameworkk"
```
Comment (YakumoChen) [inline modular_skyrat/modules/modular_implants/code/nifs.dm]: Spellcheck

File: modular_skyrat/modules/modular_implants/code/nifsofts/prop_summoner.dm [hunk 2]
```diff
@@ -87,10 +89,49 @@
...
+
+/obj/item/toy/cards/deck/tarot/nanite
+	name = "tarot deck"
+	desc = "Another piece of gaming equipment graciously donated from the Selenian Wargaming Society, these cards employ a localized field of near-invisible nanites equipped with advanced eye tracking software; to ensure the display on the cards does not allow for peeking. Additionally, over five hundred thousand variations of the standard fifty two card deck are supported, in all known forms of writing. Lastly, a collaboration with the Altspace Coven has yielded a deck of tarot cards; the witches designing it claiming that the incorporeal nature of the cards allow them a higher connectivity with fate itself. "
```
Comment (YakumoChen) [inline modular_skyrat/modules/modular_implants/code/nifsofts/prop_summoner.dm]: You copy-pasted this description but forgot that only playing card decks have 52 cards. Tarot decks have 72.

File: modular_skyrat/modules/modular_implants/code/nifsofts/prop_summoner.dm [hunk 3]
```diff
@@ -87,10 +89,49 @@
...
+
+/obj/item/toy/cards/deck/kotahi/nanite
+	name = "kotahi deck"
+	desc = "Another piece of gaming equipment graciously donated from the Selenian Wargaming Society, these cards employ a localized field of near-invisible nanites equipped with advanced eye tracking software; to ensure the display on the cards does not allow for peeking. Additionally, over five hundred thousand variations of the standard fifty two card deck are supported, in all known forms of writing. Lastly, a collaboration with the Altspace Coven has yielded a deck of tarot cards; the witches designing it claiming that the incorporeal nature of the cards allow them a higher connectivity with fate itself. "
```
Comment (YakumoChen) [inline modular_skyrat/modules/modular_implants/code/nifsofts/prop_summoner.dm]: same here but I don't know how many cards are in an uno deck.

File: modular_skyrat/modules/modular_implants/code/nifsofts/prop_summoner.dm (hunk 2 above)
Comment (softcerv) [reply modular_skyrat/modules/modular_implants/code/nifsofts/prop_summoner.dm]: this is a placeholder

File: modular_skyrat/modules/modular_implants/code/nifsofts/prop_summoner.dm (hunk 3 above)
Comment (softcerv) [reply modular_skyrat/modules/modular_implants/code/nifsofts/prop_summoner.dm]: this is a placeholder

File: modular_skyrat/modules/modular_implants/code/nifs.dm [hunk 4]
```diff
@@ -255,6 +256,73 @@
 		if(nifsoft.type == nifsoft_to_find)
 			return nifsoft
 
+///NIFSoft Remover. This is mostly here so that security and antags have a way to remove NIFSofts from someome
```
Comment (YakumoChen) [inline modular_skyrat/modules/modular_implants/code/nifs.dm:259]: make sure you add a traitor-themed child or except it from being used for illegal tech when you add it to the uplink, if you plan to, so security can't use theirs for illegal.

**Comment by Gandalf2k15** [comment]
re-open once this has been completed

**Comment by softcerv** [comment]
I had this PR open so that maintainers could give me feedback on any potential code issues with the framework code for NIFs and NIFSofts.

**Comment by YakumoChen** [comment]
This is what draft is for

**Pull request closed without merging**